File: assets/rt/Numbers.java
package rt;

import java.math.BigInteger;

public final class Numbers{
  private Numbers(){}
  public static long pow64(long base, long expBits /* unsigned */){
    long res = 1L;
    for (; expBits != 0; expBits >>>= 1){
      if ((expBits & 1L) != 0) res *= base;
      base *= base;
    }
  return res;
  }

  public static byte pow8(byte base, long expBits){
    return (byte) pow64(base, expBits);
  }

  public static long intSqrt(long x){
    if (x < 0){ throw new ArithmeticException("sqrt of negative Int"); }
    if (x <= 1){ return x; }
    long r= (long) Math.sqrt((double) x);
    for (;;){
      long q= x / r;
      long nr= (r + q) >>> 1;
      if (nr >= r){ break; }
      r = nr;
    }
    long rp1= r + 1;
    if (x / rp1 >= rp1){ r = rp1; }
    return r;
  }
  public static long natSqrt(long u){
    if (u == 0 || u == 1){ return u; }
    double du= (u & Long.MAX_VALUE) + (u < 0 ? 0x1.0p63 : 0.0);
    long r= (long) Math.sqrt(du);
    for (;;){
      long q= Long.divideUnsigned(u, r);
      long nr= (r + q) >>> 1;
      if (nr >= r){ break; }
      r = nr;
    }
    long rp1= r + 1;
    if (Long.compareUnsigned(Long.divideUnsigned(u, rp1), rp1) >= 0){ r = rp1; }
    return r;
  }
  public static byte byteSqrt(byte raw) {
    int x= Byte.toUnsignedInt(raw);
    int r= (int) Math.floor(Math.sqrt((double) x));
    return (byte) r;
  }
  public static base.Bool_0 toBool(boolean b) {
    return b ? base.True_0.$self : base.False_0.$self;
  }
}

File: assets/rt-js/Numbers.js
import { base$$False_0, base$$True_0 } from "../base/index.js";

export class rt$$Numbers {
  // // exponentiation with BigInt (like pow64 in Java)
  // static pow64(base, expBits) {
  //   base = BigInt(base);
  //   expBits = BigInt(expBits);
  //   let res = 1n;
  //   while (expBits !== 0n) {
  //     if ((expBits & 1n) !== 0n) res *= base;
  //     base *= base;
  //     expBits >>= 1n;
  //   }
  //   return res;
  // }
  //
  // // pow8: byte promoted to BigInt then back to Number (0-255 wraparound)
  // static pow8(base, expBits) {
  //   const result = Numbers.pow64(BigInt(base), BigInt(expBits));
  //   return Number(result & 0xFFn); // simulate byte overflow
  // }
  //
  // integer sqrt for signed Int (BigInt)
  static intSqrt(x) {
    x = BigInt(x);
    if (x < 0n) throw new Error("sqrt of negative Int");
    if (x <= 1n) return x;

    let r = BigInt(Math.floor(Math.sqrt(Number(x))));
    while (true) {
      const q = x / r;
      const nr = (r + q) >> 1n;
      if (nr >= r) break;
      r = nr;
    }
    const rp1 = r + 1n;
    if (x / rp1 >= rp1) r = rp1;
    return r;
  }

  // sqrt for Nat (unsigned)
  static natSqrt(u) {
    u = BigInt.asUintN(64, u);
    if (u === 0n || u === 1n) return u;

    let r = BigInt(Math.floor(Math.sqrt(Number(u))));
    while (true) {
      const q = u / r;
      const nr = (r + q) >> 1n;
      if (nr >= r) break;
      r = nr;
    }
    const rp1 = r + 1n;
    if (u / rp1 >= rp1) r = rp1;
    return r;
  }

  // sqrt for Byte
  static byteSqrt(raw) {
    const x = Number(raw) & 0xFF; // unsigned byte
    const r = Math.floor(Math.sqrt(x));
    return r & 0xFF;
  }

  // convert JS boolean to Fearless Bool object
  static toBool(b) {
    return b ? base$$True_0.$self : base$$False_0.$self;
  }

  // Overflow/Underflow:  just like clocks, numbers silently wrap
  static toInt64(x) {
    const mask64 = (1n << 64n) - 1n;
    x &= mask64;
    if (x >= (1n << 63n)) x -= (1n << 64n);
    return x;
  }

  static toNat64(x) {
    const mask64 = (1n << 64n) - 1n;
    return x & mask64;
  }

}
